[Bug]: Scripts won't work

### What happened?

Scripts dosen't work on 0958db382563b29ebc0adc1b6350bd8c06b1ea4d.

### Steps to Reproduce

1.  `git clone https://github.com/danny-avila/LibreChat.git`
2.  `docker-compose up -d`
3.  `sudo docker exec -ti LibreChat sh`
4.  `npm run create-user`

### Relevant log output

```shell
root@netherlandspower:~/LibreChat# sudo docker exec -ti LibreChat sh
/app # npm run create-user

> LibreChat@0.6.1 create-user
> node config/create-user.js

node:internal/modules/cjs/loader:1093
  throw err;
  ^

Error: Cannot find module '~/strategies/validators'
Require stack:
- /app/api/server/services/AuthService.js
- /app/config/create-user.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1090:15)
    at Module._load (node:internal/modules/cjs/loader:934:27)
    at Module.require (node:internal/modules/cjs/loader:1157:19)
    at require (node:internal/modules/helpers:119:18)
    at Object.<anonymous> (/app/api/server/services/AuthService.js:3:44)
    at Module._compile (node:internal/modules/cjs/loader:1275:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1329:10)
    at Module.load (node:internal/modules/cjs/loader:1133:32)
    at Module._load (node:internal/modules/cjs/loader:972:12)
    at Module.require (node:internal/modules/cjs/loader:1157:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/app/api/server/services/AuthService.js',
    '/app/config/create-user.js'
  ]
}

Node.js v19.9.0
```

--- a/api/app/clients/AnthropicClient.js
+++ b/api/app/clients/AnthropicClient.js
@@ -290,7 +290,7 @@ class AnthropicClient extends BaseClient {
       top_p,
       top_k,
     };
-    logger.debug('[AnthropicClient]', { requestOptions });
+    logger.debug('[AnthropicClient]', { ...requestOptions });
     const response = await client.completions.create(requestOptions);
 
     signal.addEventListener('abort', () => {

--- a/api/config/winston.js
+++ b/api/config/winston.js
@@ -2,7 +2,6 @@ const path = require('path');
 const winston = require('winston');
 require('winston-daily-rotate-file');
 const { redact, deepObjectFormat } = require('./parsers');
-const { isEnabled } = require('~/server/utils/handleText');
 
 const logDir = path.join(__dirname, '..', 'logs');
 
@@ -67,7 +66,10 @@ const transports = [
 //   );
 // }
 
-if (isEnabled && isEnabled(DEBUG_LOGGING)) {
+if (
+  (typeof DEBUG_LOGGING === 'string' && DEBUG_LOGGING?.toLowerCase() === 'true') ||
+  DEBUG_LOGGING === true
+) {
   transports.push(
     new winston.transports.DailyRotateFile({
       level: 'debug',
@@ -88,7 +90,10 @@ const consoleFormat = winston.format.combine(
   winston.format.printf((info) => `${info.timestamp} ${info.level}: ${info.message}`),
 );
 
-if (isEnabled && isEnabled(DEBUG_CONSOLE)) {
+if (
+  (typeof DEBUG_CONSOLE === 'string' && DEBUG_CONSOLE?.toLowerCase() === 'true') ||
+  DEBUG_CONSOLE === true
+) {
   transports.push(
     new winston.transports.Console({
       level: 'debug',

--- a/api/lib/db/indexSync.js
+++ b/api/lib/db/indexSync.js
@@ -65,6 +65,8 @@ async function indexSync(req, res, next) {
           logger.error('[indexSync] Trouble creating indices, try restarting the server.', err);
         }
       }, 750);
+    } else if (err.message.includes('Meilisearch not configured')) {
+      logger.info('[indexSync] Meilisearch not configured, search will be disabled.');
     } else {
       logger.error('[indexSync] error', err);
       // res.status(500).json({ error: 'Server error' });

--- a/api/package.json
+++ b/api/package.json
@@ -30,7 +30,7 @@
     "@anthropic-ai/sdk": "^0.5.4",
     "@azure/search-documents": "^12.0.0",
     "@keyv/mongo": "^2.1.8",
-    "@keyv/redis": "^2.8.0",
+    "@keyv/redis": "^2.8.1",
     "@langchain/google-genai": "^0.0.2",
     "axios": "^1.3.4",
     "bcryptjs": "^2.4.3",

--- a/api/server/controllers/AskController.js
+++ b/api/server/controllers/AskController.js
@@ -4,7 +4,7 @@ const { saveMessage, getConvoTitle, getConvo } = require('~/models');
 const { createAbortController, handleAbortError } = require('~/server/middleware');
 const { logger } = require('~/config');
 
-const AskController = async (req, res, next, initializeClient) => {
+const AskController = async (req, res, next, initializeClient, addTitle) => {
   let {
     text,
     endpointOption,
@@ -12,7 +12,9 @@ const AskController = async (req, res, next, initializeClient) => {
     parentMessageId = null,
     overrideParentMessageId = null,
   } = req.body;
+
   logger.debug('[AskController]', { text, conversationId, ...endpointOption });
+
   let metadata;
   let userMessage;
   let promptTokens;
@@ -21,7 +23,10 @@ const AskController = async (req, res, next, initializeClient) => {
   let lastSavedTimestamp = 0;
   let saveDelay = 100;
   const sender = getResponseSender({ ...endpointOption, model: endpointOption.modelOptions.model });
+  const newConvo = !conversationId;
   const user = req.user.id;
+
+  const addMetadata = (data) => (metadata = data);
 
   const getReqData = (data = {}) => {
     for (let key in data) {
@@ -50,6 +55,7 @@ const AskController = async (req, res, next, initializeClient) => {
           conversationId,
           parentMessageId: overrideParentMessageId ?? userMessageId,
           text: partialText,
+          model: endpointOption.modelOptions.model,
           unfinished: true,
           cancelled: false,
           error: false,
@@ -62,46 +68,50 @@ const AskController = async (req, res, next, initializeClient) => {
       }
     },
   });
+
+  const getAbortData = () => ({
+    sender,
+    conversationId,
+    messageId: responseMessageId,
+    parentMessageId: overrideParentMessageId ?? userMessageId,
+    text: getPartialText(),
+    userMessage,
+    promptTokens,
+  });
+
+  const { abortController, onStart } = createAbortController(req, res, getAbortData);
+
   try {
-    const addMetadata = (data) => (metadata = data);
-    const getAbortData = () => ({
-      conversationId,
-      messageId: responseMessageId,
-      sender,
-      parentMessageId: overrideParentMessageId ?? userMessageId,
-      text: getPartialText(),
-      userMessage,
-      promptTokens,
-    });
-
-    const { abortController, onStart } = createAbortController(req, res, getAbortData);
-
     const { client } = await initializeClient({ req, res, endpointOption });
-
-    let response = await client.sendMessage(text, {
-      // debug: true,
+    const messageOptions = {
       user,
-      conversationId,
       parentMessageId,
+      conversationId,
       overrideParentMessageId,
-      ...endpointOption,
+      getReqData,
+      onStart,
+      addMetadata,
+      abortController,
       onProgress: progressCallback.call(null, {
         res,
         text,
-        parentMessageId: overrideParentMessageId ?? userMessageId,
+        parentMessageId: overrideParentMessageId || userMessageId,
       }),
-      onStart,
-      getReqData,
-      addMetadata,
-      abortController,
-    });
+    };
+
+    let response = await client.sendMessage(text, messageOptions);
+
+    if (overrideParentMessageId) {
+      response.parentMessageId = overrideParentMessageId;
+    }
 
     if (metadata) {
       response = { ...response, ...metadata };
     }
 
-    if (overrideParentMessageId) {
-      response.parentMessageId = overrideParentMessageId;
+    if (client.options.attachments) {
+      userMessage.files = client.options.attachments;
+      delete userMessage.image_urls;
     }
 
     sendMessage(res, {
@@ -116,7 +126,13 @@ const AskController = async (req, res, next, initializeClient) => {
     await saveMessage({ ...response, user });
     await saveMessage(userMessage);
 
-    // TODO: add title service
+    if (addTitle && parentMessageId === '00000000-0000-0000-0000-000000000000' && newConvo) {
+      addTitle(req, {
+        text,
+        response,
+        client,
+      });
+    }
   } catch (error) {
     const partialText = getPartialText();
     handleAbortError(res, req, error, {

--- a/api/server/controllers/EditController.js
+++ b/api/server/controllers/EditController.js
@@ -15,6 +15,7 @@ const EditController = async (req, res, next, initializeClient) => {
     parentMessageId = null,
     overrideParentMessageId = null,
   } = req.body;
+
   logger.debug('[EditController]', {
     text,
     generation,
@@ -22,6 +23,7 @@ const EditController = async (req, res, next, initializeClient) => {
     conversationId,
     ...endpointOption,
   });
+
   let metadata;
   let userMessage;
   let promptTokens;
@@ -48,6 +50,7 @@ const EditController = async (req, res, next, initializeClient) => {
     generation,
     onProgress: ({ text: partialText }) => {
       const currentTimestamp = Date.now();
+
       if (currentTimestamp - lastSavedTimestamp > saveDelay) {
         lastSavedTimestamp = currentTimestamp;
         saveMessage({
@@ -56,6 +59,7 @@ const EditController = async (req, res, next, initializeClient) => {
           conversationId,
           parentMessageId: overrideParentMessageId ?? userMessageId,
           text: partialText,
+          model: endpointOption.modelOptions.model,
           unfinished: true,
           cancelled: false,
           isEdited: true,
@@ -69,19 +73,20 @@ const EditController = async (req, res, next, initializeClient) => {
       }
     },
   });
+
+  const getAbortData = () => ({
+    conversationId,
+    messageId: responseMessageId,
+    sender,
+    parentMessageId: overrideParentMessageId ?? userMessageId,
+    text: getPartialText(),
+    userMessage,
+    promptTokens,
+  });
+
+  const { abortController, onStart } = createAbortController(req, res, getAbortData);
+
   try {
-    const getAbortData = () => ({
-      conversationId,
-      messageId: responseMessageId,
-      sender,
-      parentMessageId: overrideParentMessageId ?? userMessageId,
-      text: getPartialText(),
-      userMessage,
-      promptTokens,
-    });
-
-    const { abortController, onStart } = createAbortController(req, res, getAbortData);
-
     const { client } = await initializeClient({ req, res, endpointOption });
 
     let response = await client.sendMessage(text, {
@@ -93,25 +98,22 @@ const EditController = async (req, res, next, initializeClient) => {
       parentMessageId,
       responseMessageId,
       overrideParentMessageId,
-      ...endpointOption,
-      onProgress: progressCallback.call(null, {
-        res,
-        text,
-        parentMessageId: overrideParentMessageId ?? userMessageId,
-      }),
       getReqData,
       onStart,
       addMetadata,
       abortController,
+      onProgress: progressCallback.call(null, {
+        res,
+        text,
+        parentMessageId: overrideParentMessageId || userMessageId,
+      }),
     });
 
     if (metadata) {
       response = { ...response, ...metadata };
     }
 
-    if (overrideParentMessageId) {
-      response.parentMessageId = overrideParentMessageId;
-    }
+    await saveMessage({ ...response, user });
 
     sendMessage(res, {
       title: await getConvoTitle(user, conversationId),
@@ -121,11 +123,6 @@ const EditController = async (req, res, next, initializeClient) => {
       responseMessage: response,
     });
     res.end();
-
-    await saveMessage({ ...response, user });
-    await saveMessage(userMessage);
-
-    // TODO: add title service
   } catch (error) {
     const partialText = getPartialText();
     handleAbortError(res, req, error, {

--- a/api/server/routes/ask/openAI.js
+++ b/api/server/routes/ask/openAI.js
@@ -1,160 +1,19 @@
 const express = require('express');
-const router = express.Router();
-const { sendMessage, createOnProgress } = require('~/server/utils');
-const { saveMessage, getConvoTitle, getConvo } = require('~/models');
-const { getResponseSender } = require('librechat-data-provider');
+const AskController = require('~/server/controllers/AskController');
 const { addTitle, initializeClient } = require('~/server/services/Endpoints/openAI');
 const {
   handleAbort,
-  createAbortController,
-  handleAbortError,
   setHeaders,
   validateEndpoint,
   buildEndpointOption,
 } = require('~/server/middleware');
-const { logger } = require('~/config');
+
+const router = express.Router();
 
 router.post('/abort', handleAbort());
 
-router.post('/', validateEndpoint, buildEndpointOption, setHeaders, async (req, res) => {
-  let {
-    text,
-    endpointOption,
-    conversationId,
-    parentMessageId = null,
-    overrideParentMessageId = null,
-  } = req.body;
-
-  logger.debug('[/ask/openAI]', { text, conversationId, ...endpointOption });
-
-  let metadata;
-  let userMessage;
-  let promptTokens;
-  let userMessageId;
-  let responseMessageId;
-  let lastSavedTimestamp = 0;
-  let saveDelay = 100;
-  const sender = getResponseSender({ ...endpointOption, model: endpointOption.modelOptions.model });
-  const newConvo = !conversationId;
-  const user = req.user.id;
-
-  const addMetadata = (data) => (metadata = data);
-
-  const getReqData = (data = {}) => {
-    for (let key in data) {
-      if (key === 'userMessage') {
-        userMessage = data[key];
-        userMessageId = data[key].messageId;
-      } else if (key === 'responseMessageId') {
-        responseMessageId = data[key];
-      } else if (key === 'promptTokens') {
-        promptTokens = data[key];
-      } else if (!conversationId && key === 'conversationId') {
-        conversationId = data[key];
-      }
-    }
-  };
-
-  const { onProgress: progressCallback, getPartialText } = createOnProgress({
-    onProgress: ({ text: partialText }) => {
-      const currentTimestamp = Date.now();
-
-      if (currentTimestamp - lastSavedTimestamp > saveDelay) {
-        lastSavedTimestamp = currentTimestamp;
-        saveMessage({
-          messageId: responseMessageId,
-          sender,
-          conversationId,
-          parentMessageId: overrideParentMessageId ?? userMessageId,
-          text: partialText,
-          model: endpointOption.modelOptions.model,
-          unfinished: true,
-          cancelled: false,
-          error: false,
-          user,
-        });
-      }
-
-      if (saveDelay < 500) {
-        saveDelay = 500;
-      }
-    },
-  });
-
-  const getAbortData = () => ({
-    sender,
-    conversationId,
-    messageId: responseMessageId,
-    parentMessageId: overrideParentMessageId ?? userMessageId,
-    text: getPartialText(),
-    userMessage,
-    promptTokens,
-  });
-
-  const { abortController, onStart } = createAbortController(req, res, getAbortData);
-
-  try {
-    const { client } = await initializeClient({ req, res, endpointOption });
-    const messageOptions = {
-      user,
-      parentMessageId,
-      conversationId,
-      overrideParentMessageId,
-      getReqData,
-      onStart,
-      addMetadata,
-      abortController,
-      onProgress: progressCallback.call(null, {
-        res,
-        text,
-        parentMessageId: overrideParentMessageId || userMessageId,
-      }),
-    };
-
-    let response = await client.sendMessage(text, messageOptions);
-
-    if (overrideParentMessageId) {
-      response.parentMessageId = overrideParentMessageId;
-    }
-
-    if (metadata) {
-      response = { ...response, ...metadata };
-    }
-
-    if (client.options.attachments) {
-      userMessage.files = client.options.attachments;
-      delete userMessage.image_urls;
-    }
-
-    sendMessage(res, {
-      title: await getConvoTitle(user, conversationId),
-      final: true,
-      conversation: await getConvo(user, conversationId),
-      requestMessage: userMessage,
-      responseMessage: response,
-    });
-    res.end();
-
-    await saveMessage({ ...response, user });
-    await saveMessage(userMessage);
-
-    if (parentMessageId === '00000000-0000-0000-0000-000000000000' && newConvo) {
-      addTitle(req, {
-        text,
-        response,
-        client,
-      });
-    }
-  } catch (error) {
-    const partialText = getPartialText();
-    handleAbortError(res, req, error, {
-      partialText,
-      conversationId,
-      sender,
-      messageId: responseMessageId,
-      parentMessageId: userMessageId ?? parentMessageId,
-    });
-  }
+router.post('/', validateEndpoint, buildEndpointOption, setHeaders, async (req, res, next) => {
+  await AskController(req, res, next, initializeClient, addTitle);
 });
 
 module.exports = router;

--- a/api/server/routes/edit/openAI.js
+++ b/api/server/routes/edit/openAI.js
@@ -1,150 +1,19 @@
 const express = require('express');
-const router = express.Router();
-const { getResponseSender } = require('librechat-data-provider');
+const EditController = require('~/server/controllers/EditController');
 const { initializeClient } = require('~/server/services/Endpoints/openAI');
-const { saveMessage, getConvoTitle, getConvo } = require('~/models');
-const { sendMessage, createOnProgress } = require('~/server/utils');
 const {
   handleAbort,
-  createAbortController,
-  handleAbortError,
   setHeaders,
   validateEndpoint,
   buildEndpointOption,
 } = require('~/server/middleware');
-const { logger } = require('~/config');
+
+const router = express.Router();
 
 router.post('/abort', handleAbort());
 
-router.post('/', validateEndpoint, buildEndpointOption, setHeaders, async (req, res) => {
-  let {
-    text,
-    generation,
-    endpointOption,
-    conversationId,
-    responseMessageId,
-    isContinued = false,
-    parentMessageId = null,
-    overrideParentMessageId = null,
-  } = req.body;
-
-  logger.debug('[/edit/openAI]', {
-    text,
-    generation,
-    isContinued,
-    conversationId,
-    ...endpointOption,
-  });
-
-  let metadata;
-  let userMessage;
-  let promptTokens;
-  let lastSavedTimestamp = 0;
-  let saveDelay = 100;
-  const sender = getResponseSender({ ...endpointOption, model: endpointOption.modelOptions.model });
-  const userMessageId = parentMessageId;
-  const user = req.user.id;
-
-  const addMetadata = (data) => (metadata = data);
-  const getReqData = (data = {}) => {
-    for (let key in data) {
-      if (key === 'userMessage') {
-        userMessage = data[key];
-      } else if (key === 'responseMessageId') {
-        responseMessageId = data[key];
-      } else if (key === 'promptTokens') {
-        promptTokens = data[key];
-      }
-    }
-  };
-
-  const { onProgress: progressCallback, getPartialText } = createOnProgress({
-    generation,
-    onProgress: ({ text: partialText }) => {
-      const currentTimestamp = Date.now();
-
-      if (currentTimestamp - lastSavedTimestamp > saveDelay) {
-        lastSavedTimestamp = currentTimestamp;
-        saveMessage({
-          messageId: responseMessageId,
-          sender,
-          conversationId,
-          parentMessageId: overrideParentMessageId || userMessageId,
-          text: partialText,
-          model: endpointOption.modelOptions.model,
-          unfinished: true,
-          cancelled: false,
-          isEdited: true,
-          error: false,
-          user,
-        });
-      }
-
-      if (saveDelay < 500) {
-        saveDelay = 500;
-      }
-    },
-  });
-
-  const getAbortData = () => ({
-    sender,
-    conversationId,
-    messageId: responseMessageId,
-    parentMessageId: overrideParentMessageId ?? userMessageId,
-    text: getPartialText(),
-    userMessage,
-    promptTokens,
-  });
-
-  const { abortController, onStart } = createAbortController(req, res, getAbortData);
-
-  try {
-    const { client } = await initializeClient({ req, res, endpointOption });
-
-    let response = await client.sendMessage(text, {
-      user,
-      generation,
-      isContinued,
-      isEdited: true,
-      conversationId,
-      parentMessageId,
-      responseMessageId,
-      overrideParentMessageId,
-      getReqData,
-      onStart,
-      addMetadata,
-      abortController,
-      onProgress: progressCallback.call(null, {
-        res,
-        text,
-        parentMessageId: overrideParentMessageId || userMessageId,
-      }),
-    });
-
-    if (metadata) {
-      response = { ...response, ...metadata };
-    }
-
-    await saveMessage({ ...response, user });
-
-    sendMessage(res, {
-      title: await getConvoTitle(user, conversationId),
-      final: true,
-      conversation: await getConvo(user, conversationId),
-      requestMessage: userMessage,
-      responseMessage: response,
-    });
-    res.end();
-  } catch (error) {
-    const partialText = getPartialText();
-    handleAbortError(res, req, error, {
-      partialText,
-      conversationId,
-      sender,
-      messageId: responseMessageId,
-      parentMessageId: userMessageId ?? parentMessageId,
-    });
-  }
+router.post('/', validateEndpoint, buildEndpointOption, setHeaders, async (req, res, next) => {
+  await EditController(req, res, next, initializeClient);
 });
 
 module.exports = router;

--- a/config/add-balance.js
+++ b/config/add-balance.js
@@ -1,7 +1,9 @@
-const connectDb = require('../api/lib/db/connectDb');
+const path = require('path');
+require('module-alias')({ base: path.resolve(__dirname, '..', 'api') });
 const { askQuestion, silentExit } = require('./helpers');
-const User = require('../api/models/User');
-const Transaction = require('../api/models/Transaction');
+const Transaction = require('~/models/Transaction');
+const connectDb = require('~/lib/db/connectDb');
+const User = require('~/models/User');
 
 (async () => {
   /**

--- a/config/ban-user.js
+++ b/config/ban-user.js
@@ -1,7 +1,9 @@
-const connectDb = require('../api/lib/db/connectDb');
+const path = require('path');
+require('module-alias')({ base: path.resolve(__dirname, '..', 'api') });
 const { askQuestion, silentExit } = require('./helpers');
-const banViolation = require('../api/cache/banViolation');
-const User = require('../api/models/User');
+const banViolation = require('~/cache/banViolation');
+const connectDb = require('~/lib/db/connectDb');
+const User = require('~/models/User');
 
 (async () => {
   /**

--- a/config/create-user.js
+++ b/config/create-user.js
@@ -1,7 +1,9 @@
-const connectDb = require('../api/lib/db/connectDb');
-const { registerUser } = require('../api/server/services/AuthService');
+const path = require('path');
+require('module-alias')({ base: path.resolve(__dirname, '..', 'api') });
+const { registerUser } = require('~/server/services/AuthService');
 const { askQuestion, silentExit } = require('./helpers');
-const User = require('../api/models/User');
+const connectDb = require('~/lib/db/connectDb');
+const User = require('~/models/User');
 
 (async () => {
   /**

--- a/config/loader.js
+++ b/config/loader.js
@@ -1,7 +1,7 @@
-const dotenv = require('dotenv');
-const path = require('path');
 const fs = require('fs');
+const path = require('path');
 const crypto = require('crypto');
+const dotenv = require('dotenv');
 
 /**
  * This class is responsible for loading the environment variables

--- a/config/packages.js
+++ b/config/packages.js
@@ -1,6 +1,6 @@
-const { execSync } = require('child_process');
-const path = require('path');
 const fs = require('fs');
+const path = require('path');
+const { execSync } = require('child_process');
 
 const { deleteNodeModules } = require('./helpers');
 

--- a/config/stop-backend.js
+++ b/config/stop-backend.js
@@ -1,7 +1,5 @@
-// eslint-disable-next-line
-const helpers = require('./helpers');
-const { exec } = require('child_process');
 const { promisify } = require('util');
+const { exec } = require('child_process');
 
 const isWindows = process.platform === 'win32';
 const execAsync = promisify(exec);

--- a/config/update.js
+++ b/config/update.js
@@ -1,5 +1,5 @@
-const { execSync } = require('child_process');
 const path = require('path');
+const { execSync } = require('child_process');
 const { askQuestion, isDockerRunning, deleteNodeModules, silentExit } = require('./helpers');
 
 const config = {

--- a/config/upgrade.js
+++ b/config/upgrade.js
@@ -1,8 +1,8 @@
 /**
  * Upgrade script
  */
-const dotenv = require('dotenv');
 const fs = require('fs');
+const dotenv = require('dotenv');
 const { exit } = require('process');
 
 // Suppress default warnings

--- a/packages/data-provider/.gitignore
+++ b/packages/data-provider/.gitignore
@@ -1,1 +1,2 @@
 node_modules/
+test_bundle/

--- a/packages/data-provider/package.json
+++ b/packages/data-provider/package.json
@@ -21,6 +21,7 @@
     "clean": "rimraf dist",
     "build": "npm run clean && rollup -c --silent --bundleConfigAsCjs",
     "build:watch": "rollup -c -w",
+    "rollup:api": "npx rollup -c server-rollup.config.js --bundleConfigAsCjs",
     "test": "jest --coverage --watch",
     "test:ci": "jest --coverage --ci",
     "verify": "npm run test:ci",
@@ -43,10 +44,15 @@
     "zod": "^3.22.4"
   },
   "devDependencies": {
+    "@babel/plugin-proposal-nullish-coalescing-operator": "^7.18.6",
+    "@babel/plugin-proposal-optional-chaining": "^7.21.0",
     "@babel/preset-env": "^7.21.5",
     "@babel/preset-react": "^7.18.6",
     "@babel/preset-typescript": "^7.21.0",
+    "@rollup/plugin-alias": "^5.1.0",
+    "@rollup/plugin-babel": "^6.0.4",
     "@rollup/plugin-commonjs": "^25.0.2",
+    "@rollup/plugin-json": "^6.1.0",
     "@rollup/plugin-node-resolve": "^15.1.0",
     "@rollup/plugin-replace": "^5.0.5",
     "@rollup/plugin-terser": "^0.4.4",

--- a/packages/data-provider/server-rollup.config.js
+++ b/packages/data-provider/server-rollup.config.js
@@ -1,0 +1,54 @@
+import path from 'path';
+import resolve from '@rollup/plugin-node-resolve';
+import commonjs from '@rollup/plugin-commonjs';
+import alias from '@rollup/plugin-alias';
+import json from '@rollup/plugin-json';
+import babel from '@rollup/plugin-babel';
+
+const rootPath = path.resolve(__dirname, '../../');
+const rootServerPath = path.resolve(__dirname, '../../api');
+const entryPath = path.resolve(rootPath, 'api/server/index.js');
+
+console.log('entryPath', entryPath);
+
+// Define your custom aliases here
+const customAliases = {
+  entries: [{ find: '~', replacement: rootServerPath }],
+};
+
+export default {
+  input: entryPath,
+  output: {
+    file: 'test_bundle/bundle.js',
+    format: 'cjs',
+  },
+  plugins: [
+    alias(customAliases),
+    resolve({
+      preferBuiltins: true,
+      extensions: ['.js', '.json', '.node'],
+    }),
+    commonjs(),
+    json(),
+    babel({
+      exclude: 'node_modules/**',
+      babelHelpers: 'bundled',
+      presets: [
+        ['@babel/preset-env', { targets: { node: 'current' } }],
+        '@babel/preset-typescript',
+      ],
+      plugins: [
+        '@babel/plugin-syntax-dynamic-import',
+        '@babel/plugin-proposal-nullish-coalescing-operator',
+        '@babel/plugin-proposal-optional-chaining',
+      ],
+    }),
+  ],
+  external: (id) => {
+    // More selective external function
+    if (/node_modules/.test(id)) {
+      return !id.startsWith('langchain/');
+    }
+    return false;
+  },
+};

